### DISCUSSION TITLE
GPS with arcOS

### DISCUSSION THREAD

**Discussion by chuckm2000** [kg4vdk/arcos-linux-modules#16] [General Discussion]
Hi Mike, I thought this might be worth a separate topic since others might have GPS questions in the future.

As I mentioned earlier, I have taken delivery of a very nice BU353-N5 GPS USB puck, but when I plug it in it goes nowhere.  I have tried it on arcOS, on vanilla Linux Mint, and on Ubuntu for the RaspPi4.  The symptoms are:
- Looking at the 'dmesg' output, the device is detected, and in all cases the device's red LED is on, so it's getting power.
- On arcOS it is detected but there is no Prolific driver found.  (FYI, I'm not fond of the Prolific chip, wish it was FTDI)
- On vanilla Mint, it is detected and the pl2303 driver associates it with ttyUSB0
- On RaspPi Ubuntu, it is detected and the pl2303 driver associates it with ttyUSB0.
- In all cases, after a very long wait, the device is still only showing a single red LED.
- BUT ... on both Mint and Ubuntu, xgps doesn't see any GPS strings, so it's as if the 2303 driver isn't enabled.

At this point, I'm a bit flummoxed since I'm not aware of how to get the device working on Linux. Does some sort of trigger need to be sent to wake it up? The document that came with it didn't mention a default Baud rate, and I'm not sure how to pass that along to 'gpsd' with systemctl (the default is to autobaud).  The systemctl status output shows the string that connects gpsd with ttyUSB0 but it's greyed out which usually means it isn't doing anything, so no real connection. 

Perhaps you have already solved the problem since you have been using the same GPS puck for a while.

Lots of fun with this, but it is a bit of a mystery at the moment.  There is always a solution with Linux  :-) 

73 and thanks!
Chuck. VA7UL

**Comment by kg4vdk** [comment]
I grabbed mine out of the truck, and it turns out to be a slightly different model. Mine is the BU353-S4. Mine also uses the pl2303 driver. FWIW, here's the dmesg output I see:

```shell
[  +5.493698] usb 2-3: new full-speed USB device number 10 using xhci_hcd
[  +0.165077] usb 2-3: New USB device found, idVendor=067b, idProduct=2303, bcdDevice= 4.00
[  +0.000009] usb 2-3: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[  +0.000003] usb 2-3: Product: USB-Serial Controller D
[  +0.000002] usb 2-3: Manufacturer: Prolific Technology Inc. 
[  +0.010082] pl2303 2-3:1.0: pl2303 converter detected
[  +0.030521] usb 2-3: pl2303 converter now attached to ttyUSB0
[  +0.758297] pps pps0: new PPS source usbserial0
[  +0.000043] pps pps0: source "/dev/ttyUSB0" added
``` 

How does yours differ? You mention there is no Prolific driver found. What indicates that? How long is the "long wait"? I have had some GPS modules that take forever and a day to get going if they've been moved long distances while powered off. Are you testing under unobstructed sky? Sorry if any of this sounds "is it plugged in"-ish...just don't want to overlook horses to go look for zebras.

**Comment by chuckm2000** [comment]
Well, I have a bit of egg on my face ... I suddenly realized that I have to modify the /etc/default/gpsd config file to hard code /dev/ttyUSB0 as the device, and add "-s 4800" as the Baud rate for the GlobalSat device.  It seems GPSD's autobaud doesn't always work.

I now have the GPS daemon running, and xgps works a treat. 

So far, so good!
Chuck.  VA7UL/VA7ULD

**Reply by kg4vdk** [reply]
Good deal! Glad you figured out the magic sauce for it!

**Comment by KJ5CMF** [comment]
Chuck, I picked up the same GPS device as you, and it appears I am having a similar issue you were, except Arcos does not recognize the GPS receiver(at least as indicated on the desktop by my callsign) This looks like it may be the receiver? This is from running lsusb - Bus 003 Device 004: ID 067b:23a3 Prolific Technology, Inc. USB-Serial Controller
Would you mind sharing with me how to change the baud rate? I am trying to google around and do it myself, but I am pretty inexperienced with Linux, I mainly run windows. I am not sure if this is the right area or not to edit the baud rate, but if so, it is telling me I do not have the permissions to do so.
![Screenshot from 2024-12-14 04-34-27](https://github.com/user-attachments/assets/74855d69-0c8a-44ec-bff8-e9447f7bf8fa)

**Reply by kg4vdk** [reply]
That file is owned by root, so you'll need to edit it with elevated privileges. If you prefer the GUI editor, you can use `sudo gedit /etc/default/gpsd`, or you can use a terminal based editor like `nano` with `sudo nano /etc/default/gpsd`. With either way, once edited, you'll need to restart the gpsd service using `sudo systemctl restart gpsd.service`.

Here's a nano "cheat sheet":

https://www.nano-editor.org/dist/latest/cheatsheet.html

Heads up, none of the changes will survive a reboot. But, once you figure out something that works for you, you can script your solution in a user module. You'd need to move a copy of the gpsd file with your edits into place, and restart gpsd in your script.

**Reply by KJ5CMF** [reply]
Mike, thanks again for the help! I think I am getting somewhere with the info you have provided. I have been able to get it to work as far as getting a GPS fix and have verified through cgps, YAAC, and aprs.fi. Here are the settings that worked for me. Is there a way to get it to run the script at every boot, or do I need to go back and edit after each boot? Just asking cause I am not sure, I am happy enough that I got it to work. I tried putting together a script but I am assuming I am missing something? I'll post screenshots of everything.

**Reply by KJ5CMF** [reply]
![Screenshot from 2024-12-15 18-13-45](https://github.com/user-attachments/assets/17db0ed3-3763-450d-b9e6-0ae957655f11)
![Screenshot from 2024-12-15 18-17-04](https://github.com/user-attachments/assets/e48d5796-51ac-4082-b775-924787b2ce62)
![Screenshot from 2024-12-15 18-19-11](https://github.com/user-attachments/assets/17c00411-6f77-4106-9b9e-54f4f6e8fa96)
![Screenshot from 2024-12-15 19-04-21](https://github.com/user-attachments/assets/ee9cd47b-edf5-4bc1-a67e-879f3463aada)

**Reply by kg4vdk** [reply]
To get your changes to load at Station Setup time, you need 2 components:

1. Your modified `/etc/default/gpsd` file
2. A script to put that file in place, and then restart the `gpsd.service`

It looks like you're trying to do both of those inside your `GPSFIX.sh` file. That won't work. Your `GPSFIX.sh` could look something like this:

> sudo cp /path/to/modified/gpsd /etc/default/gpsd
sudo systemctl restart gpsd.service

Where you choose to keep your modified gpsd config file is up to you. You may want to look at the example user module in arcos-linux-modules/USER/EXAMPLE, or take a look at the PARADE Community module (specifically, the directory structure regarding 'config' and line 29 of 00_PARADE.sh for copying a config into place).

A word of caution, though. If you are hard-coding `/dev/ttyUSB1`, things will not work as expected if the GPS isn't given the same `ttyUSBN` number every time. I wonder if simply adding the `-s 9600` to the shipped gpsd config might be enough for gpsd to recognize the device as a GPS. If so, that would prevent the need for hard-coding the device as something that is unpredictable.
